lib/web_ui/lib/src/engine/dom.dart

```diff
@@ -3,37 +3,85 @@
 // found in the LICENSE file.
 
 import 'dart:async';
+import 'dart:js_interop';
 import 'dart:math' as math;
 import 'dart:typed_data';
 
 import 'package:js/js.dart';
 import 'package:js/js_util.dart' as js_util;
 import 'package:meta/meta.dart';
 
+import 'browser_detection.dart';
+
 /// This file contains static interop classes for interacting with the DOM and
 /// some helpers. All of the classes in this file are named after their
 /// counterparts in the DOM. To extend any of these classes, simply add an
 /// external method to the appropriate class's extension. To add a new class,
 /// simply name the class after it's counterpart in the DOM and prefix the
 /// class name with `Dom`.
-/// NOTE: After the new static interop DOM API is released in the Dart SDK,
-/// these classes will be replaced by typedefs.
 /// NOTE: Currently, optional parameters do not behave as expected.
 /// For the time being, avoid passing optional parameters directly to JS.
-/// NOTE: Please avoid typing external JS functions with `int`, as passing ints
-/// to and from JS is not supported on all web backends.
+
+/// Conversions methods to facilitate migrating to JS types.
+///
+/// The existing behavior across the JS interop boundary involves many implicit
+/// conversions. For efficiency reasons, on JS backends we still want those
+/// implicit conversions, but on Wasm backends we need to 'shallowly' convert
+/// these types.
+///
+/// Note: Due to discrepancies between how `null`, `JSNull`, and `JSUndefined`
+/// are currently represented across web backends, these extensions should be
+/// used carefully and only on types that are known to not contains `JSNull` and
+/// `JSUndefined`.
+extension ObjectToJSAnyExtension on Object {
+  JSAny get toJSAnyShallow {
+    if (isWasm)  {
+      return toJSAnyDeep;
+    } else {
+      // TODO(joshualitt): remove this cast when we reify JS types on JS
+      // backends.
+      // ignore: unnecessary_cast
+      return this as JSAny;
+    }
+  }
+
+  JSAny get toJSAnyDeep => js_util.jsify(this) as JSAny;
+}
+
+extension JSAnyToObjectExtension on JSAny {
+  Object get toObjectShallow {
```

Comment:
For sanity sake, it does not, because we can't do better than users can do in their own code. Ideally we'd know what types flow to / from where,  and then we could write custom conversion code, perhaps writing our little JS union types? If not, then `dartify` is the answer, but I made `dartify` / `jsify` faster for native types, so it shouldn't be any worse than a giant `if` / `else if` block that we would have to use if we didn't know the JS type we were converting from.